apps/wappush/js/messagedb.js

```diff
@@ -96,35 +131,33 @@ var MessageDB = (function() {
   /**
    * Process a message executing the associated action. This deals with
    * messages with a 'delete' action by deleting all coresponding message and
-   * with new messages by storing them. Once processing is done the message is
-   * returned or null if it was discarded.
+   * with new messages by storing them.
    *
    * @param {Object} store The messages' object store.
    * @param {Object} message The message to be processed.
-   * @param {String} status The status of this message
-   *
-   * @return {String} The status of the message after processing it.
+   * @param {Boolean} [updated] True if the message is an update to an existing
+   *        one, false otherwise. Assumed false if not present.
    */
-  function mdb_processMessage(store, message, status) {
+  function mdb_processMessage(store, message, updated) {
+    updated = updated || false;
+
     /* If the message has a 'delete' action delete all messages with the
      * corresponding 'si-id' field and drop the current message. All other
      * messages should be stored. */
     if (message.action === 'delete') {
       mdb_deleteById(store, message.id);
-      status = 'discarded';
+      mdb_dispatchEvent('discard', message);
     } else {
       store.put(message);
+      mdb_dispatchEvent(updated ? 'update' : 'new', message);
     }
-
-    return status;
```

Comment:
I don't really like that processMessage dispaches the event. I think it's not its role.
Also the "updated" boolean is quite bad.

How about just returning a boolean (or a string) telling whether the message has been deleted ? And then the caller can do whatever needed to have the proper status.